x-pack/plugins/security_solution/public/detection_engine/rule_management_ui/components/rules_table/upgrade_prebuilt_rules_table/use_prebuilt_rules_upgrade_state.ts

```diff
@@ -50,6 +61,57 @@ export function usePrebuiltRulesUpgradeState(
     []
   );
 
+  // Implements concurrency control.
+  // Rule may be edited or a new prebuilt rules package version gets released.
+  // In any case current rule's `revision` or target rule's version
+  // will have higher values.
+  // Reset resolved conflicts in case of revision`s or version`s mismatch.
+  useEffect(() => {
+    for (const {
+      rule_id: ruleId,
+      current_rule: { revision: nextRevision, name },
+      target_rule: { version: nextVersion },
+    } of ruleUpgradeInfos) {
+      const cc = concurrencyControl.current[ruleId];
+      const hasNewerRevision = cc ? nextRevision > cc.revision : false;
+      const hasNewerVersion = cc ? nextVersion > cc.version : false;
+      const hasResolvedValues = Object.keys(rulesResolvedConflicts[ruleId] ?? {}).length > 0;
+
+      if (hasNewerRevision && hasResolvedValues) {
+        addWarning({
+          title: i18n.RULE_NEW_REVISION_DETECTED_WARNING,
+          text: i18n.RULE_NEW_REVISION_DETECTED_WARNING_DESCRIPTION(name),
+        });
+      }
+
+      if (hasNewerVersion && hasResolvedValues) {
+        addWarning({
+          title: i18n.RULE_NEW_VERSION_DETECTED_WARNING,
+          text: i18n.RULE_NEW_VERSION_DETECTED_WARNING_DESCRIPTION(name),
+        });
+      }
```

Comment:
There is a trade off between displaying a single but less informative toast or displaying multiple but detailed rule oriented toasts. The changes are fetched in batch via `_review` API endpoint leading to multiple toasts appearing on a screen. We have two cases

- revision change
  Probability of this case depends on users activity. It looks useful to know which rule was edited.
- version change
  I looked at prebuilt rules packages and noticed there a few rules have updates in sequential packages. Chances of that case are quite low.

Taking into account the information above I lean towards having a toast per rule. @xcrzx and @jpdjere it'd be good hear your opinion.

Reply:
This sounds good to me. Given how the code is structured and the current UX, it's highly unlikely that the screen will get flooded with toast messages.